Update the courses management overview

FrankeLukas opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Description

As discussed in meetings and on Slack when the mock-up was shared. The new overview has tiles for each course, and shows exercises split into the 3 categories "future" (not released), "current" (not due) and "past" (assessment in progress or due). Each tile also features statistics, which at the moment only include the active students, but will be expanded in the future. (E.g. with course overall score.)

### Steps for Testing

1. View the Course Management page
2. Check if everything displays correctly (e.g. no "null" or "undefined" strings in the statistics)
3. Check if the links and buttons work as intended
4. Check that as far as you can tell, that the progress bars on the left side of the exercise-specific buttons are correct
4. Make sure there is no big performance impact when using the page only for transit (i.e. the courses should load fast, while the exercises and stats can take a bit longer, allowing to access the courses fast without waiting for the other things to load)
5. The guided tour should still work.

We also suggest to test the sorting and stats of exercises:
1. Create an exercise with the release date in the future. It should be in the "Released Soon" group if the release date is within the next seven days, otherwise it should not be displayed.
2. Release the exercise it should now be in "Currently In Progress". Add a due date in the future to the exercise and it should still be in that group. While in this group the amount of participations divided by the total number of students is shown as statistic in the progress bar (in the exercise list).
3. Set an assessment due date in the future. Once the due date passes the exercise appears in "Currently In Assessment" (if **no** assessment due date is set at this point, the exercise shows in the "Past Exercises" group).
4. While in "Currently In Assessment" the number of assessments divided by the number of submissions is displayed.
5. Once the assessment due date passes the exercise is in "Past Exercises" where the average score is shown.

After participating in an exercise the "Active Students" graph should resemble that.

### Test Coverage

course-management-overview-exercise-row: 100%
course-management-overview-card: 100%
course-management-overview-statistics: 100%
course-management: 86.59%
CourseResource: 91%
CourseService: 94%
ExerciseService: 89%

### Screenshots

Every Course is displayed as one tile. Categories are listed next to the title. Exercises are grouped into "Released Soon" (release in the next 7 days), "Currently In Progress" (released, but not due), "Currently In Assessment" (due for students but not assessment due) and "Past Exercises" (assessment was over in the last 7 days). Hidden quiz exercises are not show, but visible ones are in "Released Soon" until they get set to active and move to the "Currently In Progress" group. The tiles are arranged among each other. A tile itself looks like this:
<!--[Screenshot 2021-02-23 at 21 23 36](https://user-images.githubusercontent.com/44401560/108903853-121e3c80-761e-11eb-80dc-dcc12b1b754d.png)-->
<!--![grafik](https://user-images.githubusercontent.com/72132281/109004296-ded4bf80-76a8-11eb-85ba-758d52f21ac7.png)-->
<!--![grafik](https://user-images.githubusercontent.com/72132281/111882811-e5104000-89b7-11eb-9ee4-48374a0d02d7.png)
![grafik](https://user-images.githubusercontent.com/72132281/111965952-a1702000-8af6-11eb-9b9e-a8d99c1c9472.png)
![grafik](https://user-images.githubusercontent.com/72132281/111990267-76e09000-8b13-11eb-84a5-3bf62dbb8c69.png)-->
![Screenshot 2021-03-22 at 14 34 31](https://user-images.githubusercontent.com/44401560/111998440-481ae780-8b1c-11eb-85b1-885c392195cd.png)


Tooltips are present for all buttons to be able to tell them apart.

Different groups (Released soon, Currently In Progress, etc.) provide different kinds of information shown in the progress bar. " Currently In Progress" for example, shows the current amount of participations:
<!--![grafik](https://user-images.githubusercontent.com/72132281/111882969-c9596980-89b8-11eb-9a99-629e703c37b0.png)
![Screenshot 2021-02-23 at 21 24 39](https://user-images.githubusercontent.com/44401560/108904409-bd2ef600-761e-11eb-8527-2da44a811f63.png)
![grafik](https://user-images.githubusercontent.com/72132281/111990735-da6abd80-8b13-11eb-9791-87aafb66b34f.png)-->
![Screenshot 2021-03-22 at 14 42 21](https://user-images.githubusercontent.com/44401560/111999025-ddb67700-8b1c-11eb-808c-41b224a73076.png)



Courses can also have icons, which are displayed as well (see top left):
<!--[Screenshot 2021-02-23 at 18 51 35](https://user-images.githubusercontent.com/44401560/108886040-cad98100-7608-11eb-8ea8-dbb75523d705.png)
![grafik](https://user-images.githubusercontent.com/72132281/109004598-3ffc9300-76a9-11eb-9763-a7c01ed05cff.png)
![grafik](https://user-images.githubusercontent.com/72132281/111979342-40037d80-8b05-11eb-8f43-e132415ef2f5.png)
![grafik](https://user-images.githubusercontent.com/72132281/111990650-c1620c80-8b13-11eb-92d3-d00c35d3c294.png)-->
![Screenshot 2021-03-22 at 14 26 24](https://user-images.githubusercontent.com/44401560/111997019-e4dc8580-8b1a-11eb-9593-270b25f44fa6.png)

When resizing the window, the tiles act accordingly: First, the categories and the progress bar disappears. Step by step less and less information is shown (Compare screenshot 1 and following):
<!--[Screenshot 2021-02-23 at 21 20 48](https://user-images.githubusercontent.com/44401560/108904140-6d502f00-761e-11eb-8a32-bac16ba158d5.png)-->
<!--![grafik](https://user-images.githubusercontent.com/72132281/109004416-0a57aa00-76a9-11eb-88b5-206bc7182a58.png)-->
<!--![grafik](https://user-images.githubusercontent.com/72132281/111033093-d651e800-840f-11eb-84fe-917b1fc4987f.png)
![grafik](https://user-images.githubusercontent.com/72132281/111885568-df6f2600-89c8-11eb-824a-d7c878a979fe.png)
![grafik](https://user-images.githubusercontent.com/72132281/111967384-45a69680-8af8-11eb-8ed6-f404ffd4207d.png)
![grafik](https://user-images.githubusercontent.com/72132281/111990998-1b62d200-8b14-11eb-8f45-16d8686da9ee.png)-->
![Screenshot 2021-03-22 at 14 39 27](https://user-images.githubusercontent.com/44401560/111998606-7993b300-8b1c-11eb-964e-5d183d01efd1.png)

airelawaleria reviewed (dismissed):
Wow, looks really nice :)
Worked fine for me on ts1 (2.8 sec to load the courses, 10 sec to load the exercises&statistics). Links works as expected, even without waiting for the exercises&statistics to load

marvinchimeka reviewed (approved):
tested on ts3. Beside the small loading times everything works as intended. Very nice UI btw.

simonzilker reviewed (dismissed):
Looks great! Tested on ts3, all links worked. I really like it!

JohannesWeiss reviewed (commented):
Works as intended on ts3, loading hierarchy worked out great. Nice job!

DenizShabani reviewed (approved):
Tested on TS3 and everything looks good!

JannisBittlmayer reviewed (approved):
Tested on TS1, works and looks great!
Only thing I'd like to mention is that I think it's not directly clear what the green Icon on exercises is for (see screenshot), maybe some text (like "Grading") when you hover over those might be a small suggestion
![image](https://user-images.githubusercontent.com/73840230/105038153-40c55800-5a5f-11eb-8c39-30dc990e5605.png)

EneaGore reviewed (approved):
Changes look good so far on ts1

valeriebucher reviewed (commented):
Tested on ts1 and noticed two things:
After going to 'Course Management' I got this error message:
<img width="1408" alt="Bildschirmfoto 2021-01-25 um 23 28 11" src="https://user-images.githubusercontent.com/73841445/105774587-b6e92380-5f65-11eb-9473-1dcbc6772b0c.png">

Furthermore the tiles of the courses display neither a preview of the tasks nor the statistics:
<img width="1385" alt="Bildschirmfoto 2021-01-25 um 23 30 10" src="https://user-images.githubusercontent.com/73841445/105774960-468ed200-5f66-11eb-86a0-faba8a6ae319.png">

SimonEntholzer commented on src/main/java/de/tum/in/www1/artemis/repository/TeamRepository.java:
```diff
@@ -20,6 +20,9 @@
 
     List<Team> findAllByExerciseCourseIdAndShortName(@Param("courseId") Long courseId, @Param("shortName") String shortName);
 
+    @Query(value = "select count(distinct team) from Team team where team.exercise.id = :#{#exerciseId}")
```
```suggestion
    @Query("select count(distinct team) from Team team where team.exercise.id = :#{#exerciseId}")
```
Maybe that would fix the Codacy notice. I don't know how important it is though.

SimonEntholzer commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -359,4 +360,95 @@ public void registerUserForCourse(User user, Course course) {
+     * the List of maps result contains duplicated entries. This method compares the values and returns a List<Map<String, Object>>
+     * without duplicated entries
+     *
+     * @param result the result given by the Repository call
```
It was not so easy to understand how the method does what it does. Maybe renaming `result` to something more descriptive and explaining what it exactly is could help. Is it a List of Maps of users to their active period?

SimonEntholzer commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -783,6 +784,37 @@ public void fetchParticipationsWithSubmissionsAndResultsForCourses(List<Course>
+     * @param periodIndex period index for the user activity stats
+     * @return ResponseEntity with status
+     */
+    @GetMapping(value = "/courses/stats-for-management-overview")
```
```suggestion
    @GetMapping("/courses/stats-for-management-overview")
```
Is it necessary to use `value = ` here?

SimonEntholzer commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -783,6 +784,37 @@ public void fetchParticipationsWithSubmissionsAndResultsForCourses(List<Course>
+     */
+    @GetMapping(value = "/courses/stats-for-management-overview")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<List<CourseManagementOverviewCourseDTO>> getExerciseStatsForCourseOverview(@RequestParam(value = "courseIds[]") Long[] courseIds,
```
```suggestion
    public ResponseEntity<List<CourseManagementOverviewCourseDTO>> getExerciseStatsForCourseOverview(@RequestParam("courseIds[]") Long[] courseIds,
```
Same here, maybe it is needed. I don't know honestly :)

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-card.scss:
```diff
@@ -0,0 +1,94 @@
+.card {
+    min-height: 100%;
+    max-width: 2200px;
+    .card-heading {
```
```suggestion
    
    .card-heading {
```
Codacy: Expected empty line before rule (rule-empty-line-before)

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.component.ts:
```diff
@@ -0,0 +1,81 @@
+    // TODO:
+    JSON = JSON;
+
+    getIcon(type: ExerciseType | undefined) {
```
Why should it default to file-upload when it's undefined? Is it the most general icon? :D I guess then it makes sense

FrankeLukas replied:
I'm not sure which other icon we should use as default here, I don't think it matters as this won't be undefined in practice (but we want this for type checking).

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.scss:
```diff
@@ -0,0 +1,71 @@
+    display: block ruby;
+    overflow: hidden;
+    flex-shrink: 0;
+    .exercise-icon {
```
```suggestion
    
    .exercise-icon {
```
same as above

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.scss:
```diff
@@ -0,0 +1,71 @@
+    .exercise-icon {
+        margin-top: 10px;
+    }
+    .item-title {
```
```suggestion
    
    .item-title {
```
same here

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.scss:
```diff
@@ -0,0 +1,71 @@
+        overflow: hidden;
+        text-overflow: ellipsis;
+    }
+    .item-category {
```
```suggestion
    
    .item-category {
```

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.scss:
```diff
@@ -0,0 +1,71 @@
+
+        h4 {
+            margin-bottom: 0;
+            .badge {
```
```suggestion
            
            .badge {
```

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.scss:
```diff
@@ -0,0 +1,71 @@
+            }
+        }
+    }
+    .btn {
```
```suggestion
    
    .btn {
```

SimonEntholzer commented on src/main/webapp/app/course/manage/overview/course-management-statistics.component.ts:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    private createLabels() {
+        /*const prefix = this.translateService.instant('calendar_week');
```
I guess this can be removed if it is not needed anymore?

SimonEntholzer reviewed (changes requested):
The code looks good so far :) I found a few things you could look over. Mostly you should fix the scss Codacy style issues, I guess there are some line breaks missing.

sebastiantpeter reviewed (commented):
Looks good on ts2. However, I have no active student in the overview, although I created and graded a student submission on a programming exercise. How are the active students determined?

FrankeLukas commented:
Did you use a test user? Those are not included in the statistics. (They're not normal students.) We could remove that restriction as well.

airelawaleria reviewed (commented):
Found some overlapping elements 
![image](https://user-images.githubusercontent.com/64631786/107342244-3fb4a300-6ac0-11eb-97c7-beaf6fa59331.png)
![image](https://user-images.githubusercontent.com/64631786/107342258-44795700-6ac0-11eb-8088-30bd4b368bcd.png)

DenizShabani reviewed (commented):
Shouldnt assessment box not be ticked and also the bar be in between Due and Assessment Due(since the Assessment due is in 19 days)?
![image](https://user-images.githubusercontent.com/44246548/107360394-29b1dd00-6ad6-11eb-8695-f1a9fc4ac237.png)
the same would go for Due here as well.
![image](https://user-images.githubusercontent.com/44246548/107360550-6382e380-6ad6-11eb-9ed9-e5d7f5bc6b91.png)
I also encountered the same problem as @airelawaleria here
![image](https://user-images.githubusercontent.com/44246548/107360716-96c57280-6ad6-11eb-8644-91ba2cfb5ab5.png)
In regards to the loading times, wouldnt it maybe be better to open only the stats for the courses you want to see, instead for all courses(just an idea)? So that you can navigate faster to the course you want to look up as well.

philgabr reviewed (commented):
Works well in general, visual issues seem to be fixed.

However, the participation statistics do not update for me, even after assessing them (build 95):
![image](https://user-images.githubusercontent.com/73602874/107419834-87671900-6b18-11eb-9c82-3c34d874e1a3.png)
![image](https://user-images.githubusercontent.com/73602874/107419857-8c2bcd00-6b18-11eb-9e74-3429fd9c6cef.png)

FuchsDominik commented:
> Works well in general, visual issues seem to be fixed.
> 
> However, the participation statistics do not update for me, even after assessing them (build 95):
> ![image](https://user-images.githubusercontent.com/73602874/107419834-87671900-6b18-11eb-9c82-3c34d874e1a3.png)
> ![image](https://user-images.githubusercontent.com/73602874/107419857-8c2bcd00-6b18-11eb-9e74-3429fd9c6cef.png)

@philgabr Thanks for testing. Indeed, the participations contained a bug. Should be fixed now :)

tiffca reviewed (approved):
Everything worked as expected, nice changes!

one thing, clicking on an exercise doesn't do anything for me. For example, expected is that maybe it takes us to the exercise details page. So it's clickable, but it doesn't actually do anything.
![image](https://user-images.githubusercontent.com/63906424/108049476-9aae4300-7048-11eb-9332-f1b328e6834d.png)

FrankeLukas commented:
@tiffca clicking the titles works for me. Can you try again or provide us with an error message? (If there is one in the console.)

philgabr reviewed (dismissed):
All mentioned issues should be fixed, nice new look of the course administration :+1:

Two aspects I noticed:
- Maybe a search bar would increase the usability (if there even are instructors having access to multiple courses in reality, but navigating through the courses on the TSs is quite complex e.g.)
- When using smaller browser windows, the date & participation statistics simply "disappear", I don't know if that's intended:
![image](https://user-images.githubusercontent.com/73602874/108420811-95ade700-7234-11eb-99f1-bfc63f2a3174.png)

FrankeLukas commented:
Resolved merge conflicts.

> When using smaller browser windows, the date & participation statistics simply "disappear", I don't know if that's intended:

Yeah that is intended. There is simply not enough space to display everything.

tiffca reviewed (dismissed):
Functionality still works, clicking the text on the exercise works for me

ivanchimeno commented on src/main/webapp/app/course/manage/overview/course-management-statistics.component.ts:
```diff
@@ -0,0 +1,108 @@
+            return;
+        }
+
+        this.initialStatsReceived = true;
```
Can't you just do the initialization of the initial stats in `ngOnInit`? That should only be executed once  and you can simply remove the checks

FrankeLukas replied:
Hm, yeah this can probably change now that we rewrote how this works. Thanks

ivanchimeno commented on src/main/webapp/app/course/manage/overview/course-management-overview-statistics-dto.model.ts:
```diff
@@ -0,0 +1,9 @@
+export class CourseManagementOverviewStatisticsDto {
+    public courseId: number;
+    public exerciseDTOS: CourseManagementOverviewExerciseStatisticsDTO[];
+    public activeStudents: number[];
```
Should all these variables have a `?` at the end since they can technically be undefined?

FrankeLukas replied:
The server always sends values for those, so I don't think they can be undefined?

ivanchimeno replied:
Okay I was just wondering because of this here: `course-management-overview-exercise-details-dto.model.ts` you added question marks

ivanchimeno commented on src/main/webapp/app/course/manage/course-management.component.ts:
```diff
@@ -78,19 +81,58 @@ export class CourseManagementComponent implements OnInit, OnDestroy, AfterViewIn
                 if (this.courses.find((c) => c.testCourse) !== null) {
                     this.courseSemesters[this.courseSemesters.length] = 'test';
                     this.semesterCollapsed['test'] = false;
+                    this.coursesBySemester['test'] = this.courses.filter((c) => c.testCourse);
```
You can reuse `filter` in the if statement to increase performance:

```
const testCourses = this.courses.filter(course => course.testCourse);
if (testCourses.length > 0) { 
  ...
  this.coursesBySemester["test] = testCourses
}
```

ivanchimeno reviewed (dismissed):
Wrote some comments :)

philgabr reviewed (dismissed):
Approve.

airelawaleria reviewed (commented):
after I tried resizing the window, the diagram was not resized nicely. Also after I resized again back to the whole screen, the diagramm stayed like this. After reload, the issue was resolved.
![image](https://user-images.githubusercontent.com/64631786/109071016-870f7600-76f3-11eb-96a5-5e08c48fdb67.png)

Everything else looks fine

EneaGore reviewed (commented):
Confirming issue from @airelawaleria. My issue also seems to be that when a quiz has already started and finished the Edit button takes you to a page where you can not make changes. Would it not be better to make it take you to the re-evaluate window ?

FrankeLukas commented:
> Also after I resized again back to the whole screen, the diagramm stayed like this

We also noticed that making the window bigger again doesn't adjust the diagram, but I'm not sure if we can do anything here since that is chart.js's doing as far as I understood.

Sending the user to re-evaluate instead of edit sounds like a good idea. 👍

FuchsDominik commented:
> Confirming issue from @airelawaleria. My issue also seems to be that when a quiz has already started and finished the Edit button takes you to a page where you can not make changes. Would it not be better to make it take you to the re-evaluate window ?

@EneaGore thanks for the idea, it is introduced in a9d48cb . Could you test again and see if this is what you would expect?

tiffca reviewed (dismissed):
Changes look good. Everything works as expected

valeriebucher reviewed (approved):
Looks nice. I did not find an issue.

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/repository/TeamRepository.java:
```diff
@@ -20,6 +20,9 @@
     List<Team> findAllByExerciseCourseIdAndShortName(@Param("courseId") Long courseId, @Param("shortName") String shortName);
 
+    @Query("select count(distinct team) from Team team where team.exercise.id = :#{#exerciseId}")
+    Integer getAmountByExerciseId(@Param("exerciseId") Long exerciseId);
```
```suggestion
    Integer getNumberOfTeamsForExercise(@Param("exerciseId") Long exerciseId);
```
Maybe something like this would be more clearer :)?

FuchsDominik replied:
Sounds better :)

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -271,6 +274,107 @@ public void registerUserForCourse(User user, Course course) {
+     * Fetches Course Management data from repository and returns a list of Course DTOs
+     *
+     * @param isOnlyActive Whether or not to include courses with a past endDate
+     * @return A list of Course DTOs for the course management overview
```
It seems to return a course entity and not a DTO unless I am missing something here

FuchsDominik replied:
👍

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -271,6 +274,107 @@ public void registerUserForCourse(User user, Course course) {
+     * @param isOnlyActive Whether or not to include courses with a past endDate
+     * @return A list of Course DTOs for the course management overview
+     */
+    public List<Course> getAllCoursesForOverview(Boolean isOnlyActive) {
```
```suggestion
    public List<Course> getAllCoursesForOverview(Boolean onlyConsiderActiveCourses) {
```

FuchsDominik replied:
Renamed it to `onlyActive` since this is the term we usually use in our code

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java:
```diff
@@ -90,6 +91,34 @@ where TYPE(e) = QuizExercise
+    String findStudentGroupName(@Param("courseId") long courseId);
+
+    @Query("""
+            select s.submissionDate as day, u.login as username
```
I am not saying you should do this definitely but imo make your code a lot cleaner by:

- Create a Custom Class with the attributes you want, lets call it ActiverUserDTO with a constructor that expects the day and the username. Overwrite .equals and hashcode to compare day and username
- make the query return the custom class by doing something like `SELECT DISTINCT new com.mypackage.ActiveUserDTO(s.submissionDate, u.login) from ...` **This should remove duplicates and you can remove the whole remove duplicates from map method**

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -271,6 +274,107 @@ public void registerUserForCourse(User user, Course course) {
+     * @param startDate the startDate of the period
+     * @return A List<Map<String, Object>> containing date and amount of active users in this period
+     */
+    private List<Map<String, Object>> removeDuplicatesFromMapList(List<Map<String, Object>> result, ZonedDateTime startDate) {
```
```suggestion
    private List<Map<String, Object>> removeDuplicatesActiveUserRows(List<Map<String, Object>> activeUserRows, ZonedDateTime startDate) {
```

FuchsDominik replied:
👍

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -271,6 +274,107 @@ public void registerUserForCourse(User user, Course course) {
+     * @return an array, containing the values for each bar in the graph
+     */
+    private Integer[] createResultArray(List<Map<String, Object>> outcome, ZonedDateTime endDate) {
+        Integer[] result = new Integer[4];
```
Why 4? Maybe explain a little bit what graph and what you want at each index.

FuchsDominik replied:
👍

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -382,6 +391,109 @@ private void setAssignedTeamIdForExerciseAndUser(Exercise exercise, User user) {
+            dto.setExerciseId(exerciseId);
+            dto.setExerciseTitle((String) listElement.get("title"));
+
+            if (exerciseType == QuizExercise.class) {
```
Seems to be the perfect place for a switch maybe?

FrankeLukas replied:
The solution isn't very nice, but the type is not switchable, so we can't use that.

stefanwaldhauser reviewed (dismissed):
**Wow huge improvement! Amazing work** Just some suggestion that you could maybe try?

JannisBittlmayer reviewed (approved):
Tested on TS1, works as intended

airelawaleria reviewed (approved):
looks nice :)

marvinchimeka reviewed (dismissed):
👍

JohannesWeiss reviewed (dismissed):
Tested in testing session, looks good 👍

EneaGore reviewed (dismissed):
Looks nice.

JannisBittlmayer reviewed (dismissed):
Tested with @FlorianStang, works as intended.
Only thing I noticed is the yellow button's "hitbox" is not quite the same as the button size. This is only the case for the Re-evaluate/Edit buttons (see attached video)

https://user-images.githubusercontent.com/73840230/109980270-ff7fc380-7cff-11eb-9fd6-c1033dd2504c.mp4

FlorianStang reviewed (approved):
Tested on TS1, with @JannisBittlmayer asside from the small hitbox issue we did not encouter anymore problems.

daniels98it reviewed (commented):
Tested in testing session and looks very good :rocket: 

Only bug which occurred is about the displaying of course statistics, which do not appear immediately but needs the course buttons to be hovered to be displayed.

ivohashamov reviewed (commented):
Works and looks great on ts1. There is only a small issue where the "Active Students" graph doesn't load immediately unless one hovers over one of the buttons (Edit, Scores, Submissions)

nicolasruscher reviewed (approved):
Good job, tested it in testing session, looks good! 👍

iozbeyli commented:
![Screenshot from 2021-03-22 16-19-03](https://user-images.githubusercontent.com/16825065/112013488-5bcd4a80-8b2a-11eb-908f-460c81adc348.png)

ivohashamov reviewed (dismissed):
Reapprove. Tested on ts1

FrankeLukas commented:
> 
> 
> ![Screenshot from 2021-03-22 16-19-03](https://user-images.githubusercontent.com/16825065/112013488-5bcd4a80-8b2a-11eb-908f-460c81adc348.png)

Fixed now, I moved the categories on top of each other again.

iozbeyli reviewed (dismissed):
Tested in test session, works as intended

ivohashamov reviewed (approved):
Reapprove

nicolasruscher reviewed (dismissed):
reapprove

ivohashamov reviewed (dismissed):
Reapprove

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/enumeration/QuizStatus.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Enum for defining the quiz status - analogue to the client quizStatus in the quizExercise model
+ */
+public enum QuizStatus {
```
I guess we can delete this

FrankeLukas replied:
Done.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -268,6 +271,115 @@ public void registerUserForCourse(User user, Course course) {
+     * @param onlyActive Whether or not to include courses with a past endDate
+     * @return A list of Courses for the course management overview
+     */
+    public List<Course> getAllCoursesForOverview(Boolean onlyActive) {
```
```suggestion
    public List<Course> getAllCoursesForOverview(boolean onlyActive) {
```

krusche commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -268,6 +271,115 @@ public void registerUserForCourse(User user, Course course) {
+     * The List of maps contains duplicated entries. This method compares the values and returns a List<Map<String, Object>>
+     * without duplicated entries
+     *
+     * @param activeUserRows the result given by the Repository call
```
not a good comment, please explain the data structure

krusche commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -268,6 +271,115 @@ public void registerUserForCourse(User user, Course course) {
+            }
+        }
+        List<Map<String, Object>> returnList = new ArrayList<>();
+        users.forEach((k, v) -> {
```
k, v are not good variable names

krusche commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -268,6 +271,115 @@ public void registerUserForCourse(User user, Course course) {
+     *
+     * @param outcome A List<Map<String, Object>>, containing the content which should be refactored into an array
+     * @param endDate the endDate
+     * @return an array, containing the values for each bar in the graph
```
bad java doc comment, it's unclear what you mean here
Also the method name is rather bad, what does `createResultArray` mean?

krusche commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -268,6 +271,115 @@ public void registerUserForCourse(User user, Course course) {
+     */
+    private Integer getWeekOfDate(ZonedDateTime date) {
+        LocalDate localDate = date.toLocalDate();
+        TemporalField woy = WeekFields.of(DayOfWeek.MONDAY, 4).weekOfWeekBasedYear();
```
what does woy mean?

FrankeLukas replied:
I think WeekOfDay.

FuchsDominik replied:
weekOfYear

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -537,6 +546,66 @@ private void setAssignedTeamIdForExerciseAndUser(Exercise exercise, User user) {
+     * @param exerciseIds the ids of the exercises to get the statistics for
+     * @return A list of filled <code>CourseManagementOverviewExerciseStatisticsDTO</code>
+     */
+    public List<CourseManagementOverviewExerciseStatisticsDTO> getStatisticsForCourseManagementOverview(Long courseId, Integer amountOfStudentsInCourse, List<Long> exerciseIds) {
```
the method is rather long, please split it into multiple methods, and add some code comments, so it's clear what you are doing here.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -537,6 +546,66 @@ private void setAssignedTeamIdForExerciseAndUser(Exercise exercise, User user) {
+
+        for (var exercise : exerciseRepository.getExercisesForCourseManagementOverview(courseId, sevenDaysAgo)) {
+            var exerciseId = exercise.getId();
+            var dto = new CourseManagementOverviewExerciseStatisticsDTO();
```
dto is not a good name

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -458,10 +459,22 @@ private void validateShortName(Course course) {
+     * @param onlyActive if true, only active courses will be considered in the result
+     * @return a list of courses (the user has access to)
+     */
+    @GetMapping("/courses/course-overview")
```
```suggestion
    @GetMapping("/courses/course-management-overview")
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -458,10 +459,22 @@ private void validateShortName(Course course) {
+     */
+    @GetMapping("/courses/course-overview")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public List<Course> getAllCoursesForOverview(@RequestParam(defaultValue = "false") boolean onlyActive) {
```
```suggestion
    public List<Course> getAllCoursesForManagementOverview(@RequestParam(defaultValue = "false") boolean onlyActive) {
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -458,10 +459,22 @@ private void validateShortName(Course course) {
+    @GetMapping("/courses/course-overview")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public List<Course> getAllCoursesForOverview(@RequestParam(defaultValue = "false") boolean onlyActive) {
+        return courseService.getAllCoursesForOverview(onlyActive);
```
```suggestion
        return courseService.getAllCoursesForManagementOverview(onlyActive);
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:
```diff
@@ -874,6 +887,58 @@ public void fetchParticipationsWithSubmissionsAndResultsForCourses(List<Course>
+    @GetMapping("/courses/stats-for-management-overview")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<List<CourseManagementOverviewStatisticsDTO>> getExerciseStatsForCourseOverview(@RequestParam(defaultValue = "false") boolean onlyActive) {
+        final List<CourseManagementOverviewStatisticsDTO> courseDTOS = new ArrayList<>();
```
```suggestion
        final List<CourseManagementOverviewStatisticsDTO> courseDTOs = new ArrayList<>();
```

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/dto/CourseManagementOverviewDTO.java:
```diff
@@ -0,0 +1,28 @@
+
+import de.tum.in.www1.artemis.domain.Exercise;
+
+public class CourseManagementOverviewDTO {
```
do we really need this DTO? If we send a List of exercises, those include the course anyway. I don't see an advantage to store the courseId another time

FrankeLukas replied:
We send an list of those DTOs. The course id is to be able to identify the course it belongs to.

krusche commented on src/main/webapp/app/course/manage/course-management.service.ts:
```diff
@@ -245,6 +259,44 @@ export class CourseManagementService {
+            .pipe(
+                map((res: HttpResponse<CourseManagementOverviewDto[]>) => {
+                    if (res.body) {
+                        res.body.forEach((b) => {
```
b is not a good variable name :-(

krusche commented on src/main/webapp/app/course/manage/overview/course-management-card.component.ts:
```diff
@@ -0,0 +1,85 @@
+        const inSevenDays = moment().add(7, 'days').endOf('day');
+        const sevenDaysAgo = moment().subtract(7, 'days').startOf('day');
+        this.futureExercises = exercises
+            .filter((e) => e.releaseDate && e.releaseDate > moment() && e.releaseDate <= inSevenDays)
```
e is not a good variable name

krusche commented on src/main/webapp/app/course/manage/overview/course-management-card.component.ts:
```diff
@@ -0,0 +1,85 @@
+            })
+            .slice(0, 5);
+        this.currentExercises = exercises.filter(
+            (e) => (e.releaseDate && e.releaseDate <= moment() && (!e.dueDate || e.dueDate > moment())) || (!e.releaseDate && e.dueDate && e.dueDate > moment()),
```
don't use `e`

krusche commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.component.ts:
```diff
@@ -0,0 +1,96 @@
+    icon: IconProp;
+    iconTooltip: string;
+
+    getIcon(type: ExerciseType | undefined): IconProp {
```
are you sure that this has not been implemented somewhere else? at least we already use this somewhere else

krusche commented on src/main/webapp/app/course/manage/overview/course-management-exercise-row.component.ts:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    getIconTooltip(type: ExerciseType | undefined): string {
+        switch (type) {
```
same feedback as above

krusche commented on src/main/webapp/app/course/manage/overview/course-management-overview-dto.model.ts:
```diff
@@ -0,0 +1,8 @@
+import { Exercise } from 'app/entities/exercise.model';
+
+export class CourseManagementOverviewDto {
+    public courseId: number;
```
are you sure those values always exist? is this 100% guaranteed?

FrankeLukas replied:
Yes, we always explicitly set this on the server side.

FrankeLukas replied:
We can change it to use a `?` though.

krusche commented on src/main/webapp/app/course/manage/overview/course-management-overview-statistics-dto.model.ts:
```diff
@@ -0,0 +1,9 @@
+import { CourseManagementOverviewExerciseStatisticsDTO } from 'app/course/manage/overview/course-management-overview-exercise-statistics-dto.model';
+
+export class CourseManagementOverviewStatisticsDto {
+    public courseId: number;
```
are you sure those values always exist? is this 100% guaranteed?

FrankeLukas replied:
Yes, we always explicitly set this on the server side.

krusche reviewed (changes requested):
The code still smells too much, please see my inline feedback and improve the code quality. Some issues I already mentioned several times and it's quite frustrating that you still use one letter variables :-(

derLalla reviewed (changes requested):
These are a lot of changes to review, but I am sharing my current status in the inline comments :)
